compiler/rustc_const_eval/src/interpret/cast.rs

```diff
@@ -37,6 +37,12 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 self.write_immediate(res, dest)?;
             }
 
+            PointerAddress => {
+                let src = self.read_immediate(src)?;
+                let res = self.pointer_address_cast(&src, cast_ty)?;
+                self.write_immediate(res, dest)?;
+            }
```

Comment:
So now we have two implementations of this, one in the library and one here? That is unfortunate. Also it is inconsistent to have this be a primitive but not the other direction (which would be `ptr::invalid`). This all feels rather ad-hoc.

Reply:
`ptr::addr` says this internally:
```
        // FIXME(strict_provenance_magic): I am magic and should be a compiler intrinsic.
```
If we made this an intrinsic, logically it would lower to `CastKind::PointerAddress`. Under that logic, we should have a `CastKind` for `ptr::invalid`. I don't mind implementing that.

So you are correct that this is ad-hoc, but perhaps less so than you thought. Of course I'm quite willing to do whatever the better approach is, if I could have some guidance on exactly what that is.

I'm also not opposed to changing the implementation of this pass to call `mem::transmute` or `ptr::addr`, but I can't figure out how to do that. My only idea at the moment is to turn transmute into a lang item so that I can use `Operand::function_handle(tcx.require_lang_item)`. Miri also has this `try_resolve_did` helper in it, but that seems like the wrong kind of thing to be using inside of a MIR transform.

Reply:
It does but I am not sure I agree with that comment... turns out `addr` has *exactly* the semantics we want for ptr-to-int transmutation. (That was not clear yet when Gankra wrote this comment.)

But yeah this is the usual problem with needing something in a compiler pass... we also have an `Offset` MIR operator that only exists for drop shims. I don't know a great solution.

Reply:
I am also trying to write a MIR transform that looks for occupied niches. I think that has a similar need to transmute, so I'm trying to look for one solution that will work in both transforms (I don't want to hack something together here then have the same discussion in a few months).

I have hacked together a transmute in MIR by doing AddressOf + CastKind::PtrToPtr + Deref. I think this strategy stands a good chance of breaking things though, because it takes the address of a Place (the pointer itself) that previously didn't have its address taken. And I don't see any particular reason why a function call or a new MIR operator wouldn't have the same issue.

Reply:
I created a draft PR for the niche checks, the CI should be done soon and you can look at how it breaks 2 UI tests https://github.com/rust-lang/rust/pull/104862

Reply:
Taking the address of a local should be fine? As long as it is live, that cannot introduce UB.

Maybe transmute should be added [here](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/mir/syntax/enum.NonDivergingIntrinsic.html) and then you could use it?